src/android/code-tools.ts

```diff
@@ -0,0 +1,157 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as Sentry from '@sentry/node';
+// @ts-ignore - clack is ESM and TS complains about that. It works though
+import * as clack from '@clack/prompts';
+import chalk from 'chalk';
+import {
+  sentryImport,
+  sentryImportKt,
+  testErrorSnippet,
+  testErrorSnippetKt,
+} from './templates';
+
+/**
+ * Looks in src/main/java or src/main/kotlin for the specified {@link packageName} and
+ * {@link activityName} by concatenating them. For example:
+ *
+ * src/
+ *   main/
+ *     java/ or kotlin/
+ *       my.package.name/
+ *         ui/
+ *           MainActivity.kt
+ *
+ * src/main/java can contain both .java and .kt sources, whilst src/main/kotlin only .kt
+ *
+ * @param appDir
+ * @param packageName
+ * @param activityName
+ * @returns path to the Main Activity
+ */
+export function findActivitySourceFile(
+  appDir: string,
+  packageName: string,
+  activityName: string,
+): string | undefined {
+  const javaSrcDir = path.join(appDir, 'src', 'main', 'java');
+  let possibleActivityPath;
+  const packageNameParts = packageName.split('.');
+  const activityNameParts = activityName.split('.');
+
+  if (fs.existsSync(javaSrcDir)) {
+    possibleActivityPath = `${path.join(
+      javaSrcDir,
+      ...packageNameParts,
+      ...activityNameParts,
+    )}.java`;
+    if (!fs.existsSync(possibleActivityPath)) {
+      // try kotlin if no java file
+      possibleActivityPath = possibleActivityPath.replace('.java', '.kt');
+    }
+  }
+
+  if (!possibleActivityPath || !fs.existsSync(possibleActivityPath)) {
+    const kotlinSrcDir = path.join(appDir, 'src', 'main', 'kotlin');
+    if (fs.existsSync(kotlinSrcDir)) {
+      possibleActivityPath = `${path.join(
+        kotlinSrcDir,
+        ...packageNameParts,
+        ...activityNameParts,
+      )}.kt`;
+    }
+  }
+  return possibleActivityPath;
+}
+
+/**
+ * Patches Main Activity with the test error code snippet by the specified path {@link activityFile}.
+ * Finds activity's `onCreate` method, adds the snippet and necessary imports.
+ *
+ * ```kotlin
+ * import something
+ * import something.something
+ * import io.sentry.Sentry <-- this is added by us
+ *
+ * override fun onCreate(savedInstanceState: Bundle?) {
+ *   super.onCreate(savedInstanceState)
+ *   // the snippet goes here <--
+ *   doSomething()
+ * }
+ * ```
+ *
+ * @param activityFile
+ * @returns true if successfully patched, false otherwise
+ */
+export function patchMainActivity(activityFile: string | undefined): boolean {
```

Comment:
No immediate action required but general comment RE code modifications/lookups: Using regex to check if Sentry is already used in the file or to look up/insert things is quite brittle. I think we still do this in some wizards but in others we try to parse and travers the AST. For example, to avoid flagging commented out Sentry imports as actual imports, etc. I don't think this is a necessity to change now (and frankly, I have no idea if AST parsers for Java/Kotlin exist for JS) but it's probably something to keep in mind if folks write in this in the future.

Reply:
Makes total sense! Actually, when I started with this, I wanted to modify AST for the gradle scripts, but I couldn't find any parsers for `.gradle`/`.gradle.kts` files and gave up on that idea. But I see now there are a few existing parsers at least for java/kotlin source files, so we could switch to that in the future. For now, however, I'd like to keep it simple and see how the current regex approach works out.